### DISCUSSION TITLE
Expression that can't fail, but fail

### DISCUSSION THREAD

**Discussion by jgournet** [vectordotdev/vector#19702] [Q&A] [answered]
Sorry, this is probably a silly thing I'm doing, but I don't get it:
Using "vector vrl":

Happy path:
```
$ .test = 123
123

$ .test = to_string(.test)
"123"
```

Basic failure:
```
$ .test = {"hello": "world"}
{ "hello": "world" }

$ .test = to_string(.test)

error[E103]: unhandled fallible assignment
  ┌─ :1:9
  │
1 │ .test = to_string(.test)
  │ ------- ^^^^^^^^^^^^^^^^ this expression is fallible because at least one argument's type cannot be verified to be valid
  │ │
  │ or change this to an infallible assignment:
  │ .test, err = to_string(.test)
  │
  = see documentation about error handling at https://errors.vrl.dev/#handling
  = see functions characteristics documentation at https://vrl.dev/expressions/#function-call-characteristics
  = learn more about error code 103 at https://errors.vrl.dev/103
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples


$
```
=> ok, fair enough: let me add ", err":

```
$ .test = {"hello": "world"}
{ "hello": "world" }

$ .test, err = to_string(.test)
"function call error for \"to_string\" at (13:29): unable to coerce { hello: string } into string"
```
ok, it's happy - but let me pass 123 again now:
```
$ .test = 123
123

$ .test, err = to_string(.test)

error[E104]: unnecessary error assignment
  ┌─ :1:8
  │
1 │ .test, err = to_string(.test)
  │ -----  ^^^   ---------------- because this expression can't fail
  │ │      │
  │ │      this error assignment is unnecessary
  │ use: .test = to_string(.test)
  │
  = see documentation about error handling at https://errors.vrl.dev/#handling
  = learn more about error code 104 at https://errors.vrl.dev/104
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples

$
```

sooooo : "to_string(.test)" is both faillible and can't fail ?
=> what am I doing wrong ? and how to make it work for both cases ?

Notes:
- I guess I can add a "if is_string(.test) or is_float(.test) [...etc]" before, but that sounds silly
- I tried with "to_string(.test) ?? """ with the same results

Vector version:
```
/ # vector --version
vector 0.35.0 (x86_64-unknown-linux-musl e57c0c0 2024-01-08 14:42:10.103908779)
```

**Comment by jszwedko** [comment, marked as answer]
Hi @jgournet ,

This is admittedly a bit confusing, but the fallibility of `to_string` depends on the type of the input argument, and if it is known to the compiler. In your first case, the compiler knows it is an object and so `to_string` could fail (in this case it'll always fail). In your second example it knows the input is an integer and so will never fail. Does that make sense?

**Reply by jgournet** [reply]
Thanks for saying it is confusing - I was not sure if it was me :)
Yes, it makes some sense. 
As a side note, I was using vector vrl to test my syntax and outputs before using those rules in our config file - this is why it confused me, as when running rules against our logs, I do not control the input (actually, those rules are to control the output, whatever is sent as input). Anyway, the result did confuse me, as I had thus no idea what to write in the rules :)

Thanks for your help !
